File: snippets/security/authentication/password_hashing_bcrypt.py
"""
Password Hashing with bcrypt
Production-ready password hashing and verification
"""
import bcrypt
from typing import Optional


class BcryptPasswordManager:
    """Secure password hashing using bcrypt"""

    def __init__(self, rounds: int = 12):
        """
        Initialize password manager

        Args:
            rounds: Cost factor (4-31, default 12)
                   Higher = more secure but slower
                   12 = ~0.3s, 14 = ~1.2s
        """
        if rounds < 4 or rounds > 31:
            raise ValueError("Rounds must be between 4 and 31")
        self.rounds = rounds

    def hash_password(self, password: str) -> str:
        """
        Hash a password

        Args:
            password: Plain text password

        Returns:
            Hashed password (includes salt)
        """
        # Convert password to bytes
        password_bytes = password.encode('utf-8')

        # Generate salt and hash
        salt = bcrypt.gensalt(rounds=self.rounds)
        hashed = bcrypt.hashpw(password_bytes, salt)

        # Return as string
        return hashed.decode('utf-8')

    def verify_password(self, password: str, hashed_password: str) -> bool:
        """
        Verify a password against its hash

        Args:
            password: Plain text password to verify
            hashed_password: Stored hash

        Returns:
            True if password matches
        """
        try:
            password_bytes = password.encode('utf-8')
            hashed_bytes = hashed_password.encode('utf-8')

            return bcrypt.checkpw(password_bytes, hashed_bytes)
        except Exception as e:
            print(f"Verification error: {e}")
            return False

    def needs_rehash(self, hashed_password: str) -> bool:
        """
        Check if password needs rehashing (cost factor changed)

        Args:
            hashed_password: Stored hash

        Returns:
            True if rehash needed
        """
        try:
            # Extract current rounds from hash
            hashed_bytes = hashed_password.encode('utf-8')
            # bcrypt hash format: $2b$rounds$salt+hash
            parts = hashed_bytes.split(b'$')
            if len(parts) >= 3:
                current_rounds = int(parts[2])
                return current_rounds != self.rounds
            return True
        except:
            return True

    def update_password(
        self,
        password: str,
        old_hash: str
    ) -> Optional[str]:
        """
        Update password hash if needed

        Args:
            password: Plain text password
            old_hash: Current hash

        Returns:
            New hash if update needed, None otherwise
        """
        if self.verify_password(password, old_hash):
            if self.needs_rehash(old_hash):
                return self.hash_password(password)
        return None


class PasswordStrengthValidator:
    """Validate password strength"""

    @staticmethod
    def check_strength(password: str) -> dict:
        """
        Check password strength

        Args:
            password: Password to check

        Returns:
            Dictionary with strength metrics
        """
        import re

        strength = {
            "length": len(password),
            "has_lowercase": bool(re.search(r'[a-z]', password)),
            "has_uppercase": bool(re.search(r'[A-Z]', password)),
            "has_digit": bool(re.search(r'\d', password)),
            "has_special": bool(re.search(r'[!@#$%^&*(),.?":{}|<>]', password)),
            "score": 0,
            "is_strong": False
        }

        # Calculate score
        if strength["length"] >= 8:
            strength["score"] += 1
        if strength["length"] >= 12:
            strength["score"] += 1
        if strength["length"] >= 16:
            strength["score"] += 1

        strength["score"] += sum([
            strength["has_lowercase"],
            strength["has_uppercase"],
            strength["has_digit"],
            strength["has_special"]
        ])

        # Strong password: length >= 12 and at least 3 character types
        char_types = sum([
            strength["has_lowercase"],
            strength["has_uppercase"],
            strength["has_digit"],
            strength["has_special"]
        ])

        strength["is_strong"] = strength["length"] >= 12 and char_types >= 3

        return strength

    @staticmethod
    def validate_password(password: str, min_length: int = 12) -> tuple:
        """
        Validate password meets requirements

        Args:
            password: Password to validate
            min_length: Minimum length required

        Returns:
            Tuple of (is_valid, error_message)
        """
        if len(password) < min_length:
            return False, f"Password must be at least {min_length} characters"

        strength = PasswordStrengthValidator.check_strength(password)

        if not strength["has_uppercase"]:
            return False, "Password must contain uppercase letters"

        if not strength["has_lowercase"]:
            return False, "Password must contain lowercase letters"

        if not strength["has_digit"]:
            return False, "Password must contain digits"

        if not strength["has_special"]:
            return False, "Password must contain special characters"

        return True, "Password is valid"


# Example usage
if __name__ == "__main__":
    # Initialize password manager
    pwd_manager = BcryptPasswordManager(rounds=12)

    # Hash a password
    password = "SecureP@ssw0rd123"
    hashed = pwd_manager.hash_password(password)
    print(f"Original: {password}")
    print(f"Hashed: {hashed}")

    # Verify password
    is_valid = pwd_manager.verify_password(password, hashed)
    print(f"Password valid: {is_valid}")

    # Verify wrong password
    is_valid = pwd_manager.verify_password("WrongPassword", hashed)
    print(f"Wrong password valid: {is_valid}")

    # Check password strength
    validator = PasswordStrengthValidator()
    strength = validator.check_strength(password)
    print(f"\nPassword strength: {strength}")

    # Validate password
    is_valid, message = validator.validate_password(password)
    print(f"Validation: {is_valid} - {message}")

    # Test weak password
    weak_password = "weak"
    is_valid, message = validator.validate_password(weak_password)
    print(f"Weak password validation: {is_valid} - {message}")

File: snippets/security/authentication/password_hashing_bcrypt.ts
/**
 * Password Hashing with bcrypt (TypeScript)
 * Production-ready password hashing and verification
 */
import * as bcrypt from 'bcrypt';

interface PasswordStrength {
  length: number;
  hasLowercase: boolean;
  hasUppercase: boolean;
  hasDigit: boolean;
  hasSpecial: boolean;
  score: number;
  isStrong: boolean;
}

export class BcryptPasswordManager {
  private rounds: number;

  /**
   * Initialize password manager
   * @param rounds - Cost factor (4-31, default 12)
   */
  constructor(rounds: number = 12) {
    if (rounds < 4 || rounds > 31) {
      throw new Error('Rounds must be between 4 and 31');
    }
    this.rounds = rounds;
  }

  /**
   * Hash a password
   * @param password - Plain text password
   * @returns Hashed password (includes salt)
   */
  async hashPassword(password: string): Promise<string> {
    const salt = await bcrypt.genSalt(this.rounds);
    const hashed = await bcrypt.hash(password, salt);
    return hashed;
  }

  /**
   * Verify a password against its hash
   * @param password - Plain text password to verify
   * @param hashedPassword - Stored hash
   * @returns True if password matches
   */
  async verifyPassword(
    password: string,
    hashedPassword: string
  ): Promise<boolean> {
    try {
      return await bcrypt.compare(password, hashedPassword);
    } catch (error) {
      console.error('Verification error:', error);
      return false;
    }
  }

  /**
   * Check if password needs rehashing
   * @param hashedPassword - Stored hash
   * @returns True if rehash needed
   */
  needsRehash(hashedPassword: string): boolean {
    try {
      // Extract rounds from hash ($2b$rounds$...)
      const parts = hashedPassword.split('$');
      if (parts.length >= 3) {
        const currentRounds = parseInt(parts[2], 10);
        return currentRounds !== this.rounds;
      }
      return true;
    } catch {
      return true;
    }
  }

  /**
   * Update password hash if needed
   * @param password - Plain text password
   * @param oldHash - Current hash
   * @returns New hash if update needed, null otherwise
   */
  async updatePassword(
    password: string,
    oldHash: string
  ): Promise<string | null> {
    const isValid = await this.verifyPassword(password, oldHash);

    if (isValid && this.needsRehash(oldHash)) {
      return await this.hashPassword(password);
    }

    return null;
  }
}

export class PasswordStrengthValidator {
  /**
   * Check password strength
   * @param password - Password to check
   * @returns Strength metrics
   */
  static checkStrength(password: string): PasswordStrength {
    const strength: PasswordStrength = {
      length: password.length,
      hasLowercase: /[a-z]/.test(password),
      hasUppercase: /[A-Z]/.test(password),
      hasDigit: /\d/.test(password),
      hasSpecial: /[!@#$%^&*(),.?":{}|<>]/.test(password),
      score: 0,
      isStrong: false,
    };

    // Calculate score
    if (strength.length >= 8) strength.score += 1;
    if (strength.length >= 12) strength.score += 1;
    if (strength.length >= 16) strength.score += 1;

    strength.score +=
      Number(strength.hasLowercase) +
      Number(strength.hasUppercase) +
      Number(strength.hasDigit) +
      Number(strength.hasSpecial);

    // Strong password: length >= 12 and at least 3 character types
    const charTypes =
      Number(strength.hasLowercase) +
      Number(strength.hasUppercase) +
      Number(strength.hasDigit) +
      Number(strength.hasSpecial);

    strength.isStrong = strength.length >= 12 && charTypes >= 3;

    return strength;
  }

  /**
   * Validate password meets requirements
   * @param password - Password to validate
   * @param minLength - Minimum length required
   * @returns Validation result
   */
  static validatePassword(
    password: string,
    minLength: number = 12
  ): { isValid: boolean; errors: string[] } {
    const errors: string[] = [];

    if (password.length < minLength) {
      errors.push(`Password must be at least ${minLength} characters`);
    }

    const strength = this.checkStrength(password);

    if (!strength.hasUppercase) {
      errors.push('Password must contain uppercase letters');
    }

    if (!strength.hasLowercase) {
      errors.push('Password must contain lowercase letters');
    }

    if (!strength.hasDigit) {
      errors.push('Password must contain digits');
    }

    if (!strength.hasSpecial) {
      errors.push('Password must contain special characters');
    }

    return {
      isValid: errors.length === 0,
      errors,
    };
  }

  /**
   * Check if password is commonly used
   * @param password - Password to check
   * @returns True if common
   */
  static isCommonPassword(password: string): boolean {
    const commonPasswords = [
      'password',
      '123456',
      'qwerty',
      'admin',
      'letmein',
      'welcome',
      'monkey',
      'dragon',
      'master',
      'sunshine',
    ];

    return commonPasswords.includes(password.toLowerCase());
  }
}

// Example usage
if (require.main === module) {
  (async () => {
    const pwdManager = new BcryptPasswordManager(12);

    // Hash a password
    const password = 'SecureP@ssw0rd123';
    console.log(`Original: ${password}`);

    const hashed = await pwdManager.hashPassword(password);
    console.log(`Hashed: ${hashed}`);

    // Verify password
    const isValid = await pwdManager.verifyPassword(password, hashed);
    console.log(`Password valid: ${isValid}`);

    // Verify wrong password
    const isInvalid = await pwdManager.verifyPassword('WrongPassword', hashed);
    console.log(`Wrong password valid: ${isInvalid}`);

    // Check password strength
    const strength = PasswordStrengthValidator.checkStrength(password);
    console.log('\nPassword strength:', strength);

    // Validate password
    const validation = PasswordStrengthValidator.validatePassword(password);
    console.log(`Validation: ${validation.isValid}`);
    if (validation.errors.length > 0) {
      console.log('Errors:', validation.errors);
    }

    // Test weak password
    const weakPassword = 'weak';
    const weakValidation =
      PasswordStrengthValidator.validatePassword(weakPassword);
    console.log(`\nWeak password validation: ${weakValidation.isValid}`);
    console.log('Errors:', weakValidation.errors);
  })();
}
